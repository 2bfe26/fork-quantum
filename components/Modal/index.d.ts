import { ReactNode, TouchEvent, MouseEventHandler } from 'react';

export interface ModalProps {
  children?: ReactNode[] | ReactNode;
  onClick?: (e?: MouseEvent | TouchEvent) => void;
  onClose?: MouseEventHandler<HTMLButtonElement>;
  closeButtonAriaLabel?: string;
  theme?: {
    breakpoints?: {};
    colors?: {};
    spacing?: {};
    components?: {
      button?: {};
    };
  };
}

  interface ModalHeaderProps {
    theme?: {
      spacing?: {};
    };
  }

  interface ModalTitleProps {
    small?: boolean;
    theme?: { 
      baseFontSize?: {}; 
    };
}


interface ModalContentProps {
  theme?: { 
    baseFontSize?: number;
     spacing?: {} 
    };
}

interface ModalFooterProps {
  theme?: {
    spacing?: {};
  };
}

interface CompoundedModal extends React.FunctionComponent<ModalProps> {
  Header: React.FunctionComponent<ModalHeaderProps>;
  Title: React.FunctionComponent<ModalTitleProps>;
  HeaderText: React.FunctionComponent;
  Content: React.FunctionComponent<ModalContentProps>;
  Footer: React.FunctionComponent<ModalFooterProps>;
}

const Modal: CompoundedModal = () => null;

export default Modal;
