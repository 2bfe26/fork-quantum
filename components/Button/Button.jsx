import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import SocialButtons from './SocialButtons';
import { SIZES } from '../Grid/sub-components/shared/grid-config';
import theme from '../shared/theme';
import skins from './skins';
import Icon from '../Icon/Icon';

const fontSize = ({ size }) => {
  const sizes = {
    xsmall: '12px',
    small: '12px',
    medium: '16px',
    big: '20px',
    xbig: '24px',
  };
  return `font-size: ${sizes[size] || sizes.medium};`;
};

const iconSize = ({ size }) => {
  const sizes = {
    normal: '24px',
  };

  return `font-size: ${sizes[size] || sizes.normal};`;
};

const iconMargin = ({ size }) => {
  const margins = {
    normal: '-5px 5px -7px -5px',
  };

  return `margin: ${margins[size] || margins.normal};`;
};

const padding = ({ size }) => {
  const paddings = {
    xsmall: '4px 12px',
    small: '8px 12px',
    medium: '10px 12px',
    big: '11px 16px',
    xbig: '13px 16px',
  };

  return `padding: ${paddings[size] || paddings.normal};`;
};

const ButtonIcon = styled(Icon)`
  ${iconMargin};
`;

const StyledButton = styled.button`
  border-radius: ${theme.sizes.radius};
  font-weight: bold;
  text-align: center;
  text-decoration: ${props => (props.link ? 'underline' : 'none')};

  ${fontSize}
  ${padding}
  ${theme.mixins.transition()};

  ${ButtonIcon} {
    ${iconSize}
  }

  ${props =>
    props.full &&
    `
    width: 100%;
  `}

  ${props =>
    props.center &&
    `
    display: block;
    margin-left: auto;
    margin-right: auto;
  `}

  ${props => `
    cursor: ${props.disabled ? 'not-allowed' : 'pointer'};
  `}

  ${props => {
    const { unselected, selected, disabled, shadow } = skins(props);

    return `
      background-color: ${
        props.disabled ? disabled.background : unselected.background
      };
      border: 1px solid ${props.disabled ? disabled.border : unselected.border};
      color: ${props.disabled ? disabled.color : unselected.color};
      font-weight: ${
        props.disabled ? disabled.fontWeight : unselected.fontWeight
      };

      ${shadow ? theme.mixins.shadow() : undefined};

      &:active {
        ${shadow && theme.mixins.shadow(2)};
        background-color: ${selected.background};
        border-color: ${selected.border};
        color: ${selected.color};
      }
    `;
  }}

  @media (max-width: ${SIZES.phone}px) {
    margin-left: 0;
    margin-right: 0;
    width: 100%;
  }
`;

class Button extends React.Component {
  static Facebook = SocialButtons.Facebook;

  static Google = SocialButtons.Google;

  render() {
    const { children, icon, size, ...rest } = this.props;

    return (
      <StyledButton {...rest} size={size}>
        {icon && <ButtonIcon size={size} name={icon} />}
        {children}
      </StyledButton>
    );
  }
}

Button.defaultProps = {
  center: false,
  disabled: false,
  full: false,
  icon: '',
  size: 'medium',
  skin: 'primary',
  type: 'button',
  children: 'Catho',
  onClick: () => {},
};

Button.propTypes = {
  center: PropTypes.bool,
  disabled: PropTypes.bool,
  full: PropTypes.bool,
  /** Icon name. The full catalogue can be found
   * [here](/?selectedKind=1.%20Foundation&selectedStory=Icons) */
  icon: PropTypes.string,
  size: PropTypes.oneOf(['xsmall', 'small', 'medium', 'big', 'xbig']),
  skin: PropTypes.oneOf(['primary', 'secondary', 'action']),
  type: PropTypes.oneOf(['button', 'reset', 'submit']),
  children: PropTypes.node,
  onClick: PropTypes.func,
};

export default Button;
