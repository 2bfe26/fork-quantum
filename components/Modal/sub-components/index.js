import Content from './Content';
import Footer from './Footer';
import Header from './Header';

export { Footer, Content, Header };
