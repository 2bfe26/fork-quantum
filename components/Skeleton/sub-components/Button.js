import styled from 'styled-components';
import { Rect } from './Rect';

export const Button = styled(Rect)`
  border-radius: 4px;

  height: ${({ size, theme: { spacing } }) => {
    const sizes = {
      xsmall: spacing.large,
      small: spacing.xlarge,
      medium: spacing.xxlarge,
      large: spacing.xxxlarge,
    };

    return `${sizes[size] || sizes.medium}px`;
  }};
`;
