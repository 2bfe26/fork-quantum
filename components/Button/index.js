import { Button, ButtonSimple } from './Button';

export { Button, ButtonSimple };
