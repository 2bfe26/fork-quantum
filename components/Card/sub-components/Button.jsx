import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {default as OriginalButton} from '../../Button';

const ButtonWrapper = styled(OriginalButton)`
  background-color: transparent;
  padding: 0;
  margin: 0 10px 0 0;
  box-shadow: none;
  font-family: oxygen;
  text-transform: uppercase;
  font-size: 12px;

  &:active {
    box-shadow: none;
    background-color: transparent;
    border: 1px solid transparent;
  }
`;

const Button = props => <ButtonWrapper {...props} />;

export default Button;