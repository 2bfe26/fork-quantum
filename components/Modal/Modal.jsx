import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Colors from '../Colors';
import Card from '../Card';
import Button from '../Button';
import { query } from '../Grid/sub-components/shared';
import { Content, Header, HeaderText, Title, Footer } from './sub-components';
import { theme as defaultTheme } from '../shared';

function getBreakpoint({ theme: { breakpoints } }) {
  const sizes = {
    xsmall: '90%',
    small: '400px',
    medium: '600px',
    large: '800px',
  };
  return Object.entries(sizes).map(
    ([breakpoint, value]) => query(breakpoints)[breakpoint]`width: ${value};`,
  );
}

const ModalCard = styled(Card)`
  ${Card.Header} {
    padding-right: 56px;
  }

  ${getBreakpoint}
`;

const CloseIcon = styled(Button.Icon).attrs({
  icon: 'close',
})`
  position: absolute;
  top: 16px;
  right: 16px;
`;

CloseIcon.displayName = 'CloseIcon';

const ModalWrapper = styled.div`
  align-items: center;
  background-color: ${Colors.SHADOW[50]};
  display: flex;
  height: 100vh;
  justify-content: center;
  position: fixed;
  top: 0;
  width: 100vw;
`;

ModalWrapper.displayName = 'ModalWrapper';

class Modal extends React.Component {
  static Header = Header;

  static HeaderText = HeaderText;

  static Content = Content;

  static Title = Title;

  static Footer = Footer;

  constructor(props) {
    super(props);

    this.modalWrapperRef = React.createRef();
    this.modalOverlay = document.createElement('section');
    this.focusableElements = [];
    this.focusedElementBeforeOpen = document.activeElement;
    this.firstFocusableElement = null;
    this.lastFocusableElement = null;
  }

  componentDidMount() {
    const { body } = document;

    body.appendChild(this.modalOverlay);

    this.focusableElements = this.modalOverlay.querySelectorAll(
      `a[href],
      area[href],
      input:not([disabled]),
      select:not([disabled]),
      textarea:not([disabled]),
      button:not([disabled]),
      [tabindex="0"]`,
    );
    this.firstFocusableElement = this.focusableElements[0]; // eslint-disable-line
    this.lastFocusableElement = this.focusableElements[
      this.focusableElements.length - 1
    ];
    this.firstFocusableElement.focus();

    window.addEventListener('keydown', this.handleKeyDown);
    window.addEventListener('keydown', this.handleEscKey);
  }

  componentWillUnmount() {
    const { body } = document;
    this.focusedElementBeforeOpen.focus();

    body.removeChild(this.modalOverlay);
    window.removeEventListener('keydown', this.handleKeyDown);
    window.removeEventListener('keydown', this.handleEscKey);
  }

  handleClickOutside = ({ target }) => {
    const { onClose } = this.props;
    const { current: modalRef } = this.modalWrapperRef;

    if (target === modalRef) {
      onClose();
    }
  };

  handleEscKey = ({ key }) => {
    const { onClose } = this.props;

    if (key === 'Escape') {
      onClose();
    }
  };

  handleBackwardTab = e => {
    if (document.activeElement === this.firstFocusableElement) {
      e.preventDefault();
      this.lastFocusableElement.focus();
    }
  };

  handleFowardTab = e => {
    if (document.activeElement === this.lastFocusableElement) {
      e.preventDefault();
      this.firstFocusableElement.focus();
    }
  };

  handleKeyDown = e => {
    if (e.key === 'Tab') {
      if (this.focusableElements.length === 1) {
        return e.preventDefault();
      }

      if (e.shiftKey) {
        return this.handleBackwardTab(e);
      }

      return this.handleFowardTab(e);
    }

    return false;
  };

  render() {
    const {
      children,
      onClose,
      closeButtonAriaLabel,
      theme,
      ...rest
    } = this.props;

    return ReactDOM.createPortal(
      <ModalWrapper
        onClick={this.handleClickOutside}
        ref={this.modalWrapperRef}
        role="dialog"
        {...rest}
      >
        <ModalCard theme={theme}>
          {children}
          <CloseIcon onClick={onClose} aria-label={closeButtonAriaLabel} />
        </ModalCard>
      </ModalWrapper>,
      this.modalOverlay,
    );
  }
}

Modal.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  /** Function to be called when close icon is clicked. */
  onClose: PropTypes.func,
  /** aria-label property value for the close button icon. */
  closeButtonAriaLabel: PropTypes.string,
  theme: PropTypes.shape({
    breakpoints: PropTypes.object,
  }),
};

Modal.defaultProps = {
  children: undefined,
  onClose: () => {},
  closeButtonAriaLabel: 'close dialog',
  theme: defaultTheme,
};

export default Modal;
