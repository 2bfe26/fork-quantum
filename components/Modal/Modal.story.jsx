import React from 'react';
import { storiesOf } from '@storybook/react';
import Heading from '../../.storybook/components/Heading';
import AutoPropsApi from '../../.storybook/components/AutoPropsApi';
import CodeExample from '../../.storybook/components/CodeExample';
import { TabbedView, Tab } from '../../.storybook/components/TabbedView';
import Button from '../Button';
import Modal from './Modal';

const stories = storiesOf('5. Modals', module);

class ModalStory extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showModal: false,
    };
  }

  handleOpen = () => this.setState({ showModal: !this.state.showModal });

  render() {
    return (
      <Heading atom title="Modal">
        <TabbedView>
          <Tab title="Usage">
            <Button skin="primary" onClick={this.handleOpen}>
              {this.state.showModal ? 'Close' : 'Open'} Modal
            </Button>

            { this.state.showModal &&
              (
                <Modal
                  title="Example Title"
                  closeModal={this.handleOpen}
                >
                  Example Message
                </Modal>
              )
            }

            <CodeExample component={<Modal />} />
          </Tab>

          <Tab title="API">
            <AutoPropsApi component={Modal} />
          </Tab>
        </TabbedView>
      </Heading>
    );
  }
}

stories
  .add('Basic', () => (
    <ModalStory />
  ));
