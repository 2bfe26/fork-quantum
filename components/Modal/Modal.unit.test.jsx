import React from 'react';
import { shallow } from 'enzyme';
import Modal from './Modal';
import Button from '../Button';

describe('Radio component', () => {
  const component = (
    <Modal opened trigger={<Button>Open modal</Button>}>
      <Modal.Header>Title</Modal.Header>
      <Modal.Content>Example Content</Modal.Content>
      <Modal.Footer>
        <Button skin="primary">Cancel</Button>
        <Button>OK</Button>
      </Modal.Footer>
    </Modal>
  );

  it('Should match the snapshot', () => {
    expect(shallow(component)).toMatchSnapshot();
  });

  it('should open when click trigger element', () => {
    const modal = shallow(
      <Modal trigger={<Button>Open modal</Button>}>
        <Modal.Header>Title</Modal.Header>
        <Modal.Content>Example Content</Modal.Content>
        <Modal.Footer>
          <Button skin="primary">Cancel</Button>
          <Button>OK</Button>
        </Modal.Footer>
      </Modal>,
    );
    modal.find(Button).simulate('click');
    expect(modal.state('opened')).toBe(true);
  });

  it('should close when click on overlay', () => {
    const modal = shallow(
      <Modal opened closeOnOverlayClick trigger={<Button>Open modal</Button>}>
        <Modal.Header>Title</Modal.Header>
        <Modal.Content>Example Content</Modal.Content>
        <Modal.Footer>
          <Button skin="primary">Cancel</Button>
          <Button>OK</Button>
        </Modal.Footer>
      </Modal>,
    );
    expect(modal.state('opened')).toBe(true);
    modal.find('Overlay').simulate('click', {});
    expect(modal.state('opened')).toBe(false);
  });
});
