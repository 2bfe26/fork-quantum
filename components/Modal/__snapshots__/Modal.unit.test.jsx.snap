// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio component Should match the snapshot 1`] = `"<button class=\\"Button__StyledButton-jJDkJK YtayR\\" type=\\"button\\">Open modal</button>"`;
